src/tokenizer/index.ts

```diff
@@ -87,13 +90,25 @@ export function tokenize(
     .map(token => normalizeToken(token, language, tokenizerConfig!))
     .filter(Boolean);
 
+  const tokensWithFrequency = [];
   const trimTokens = trim(tokens);
 
+  if (frequency) {
+    for (let i = 0; i < trimTokens.length; i++) {
+      const token = trimTokens[i];
+      const frequency = getTokenFrequency(token, trimTokens);
+      // @todo: replace `${token}:${frequency}` with a better form
+      tokensWithFrequency.push(`${token}:${frequency}`);
```

Comment:
I don't like this either, I'm open to suggestions

Reply:
To reduce memory usage, are you using a string rather than an object or a pair?

Reply:
Here I’m using a string to leave the tokenizer backward compatible, but I can always change this to plain Object

Reply:
ouch sorry, I missed the point 🙏

Reply:
Just my two cents, but this does not look good to me for several reason, it's conflating different responsibilities, introducing a data structure in disguise as a string, and also hidden inefficiencies.

First of all, calculating the term frequency should not be a concern of the tokenize: its responsibility is merely to split a string into tokens, nothing more.

Moreover, as you note, using a string separated by `:` does not sound like a good idea. Your instinct is right: you are representing a data structure as a string. While it may save some memory, it makes your solution more fragile and obfuscated. For the sake of backward compatibility, you are also binding to some arbitrary choices: you won't be able to support tokens that contain `:` (not possible now, but should that be a design choice?).

The `getTokenFrequency` function is also introducing an inefficiency: for each token, you go through all tokens, in order to count the frequency, which is a quadratic complexity operation. That's even repeated for every repetition of the token. Imagine a string composed of the same word 1000 times, separated by space: this would repeat the calculation for each repetition, performing 1.000.000 comparisons just to count the same term over and over.

A better architecture is to let the tokenizer tokenize, with no frequency counting, and have a clear pipeline of `tokenize -> normalize -> add tokens to index`. Downstream of the tokenizer, the part of the pipeline that receives the tokens from the tokenizer and adds them to the inverted index is the one responsible for keeping the term frequencies. It would simply be called once for each token, keep a frequency counter for each token in the inverted index, and increment it for each token added. That's better separation of responsibility, and a linear time operation instead of quadratic.

Reply:
@lucaong I totally agree and your points are the real reason why this PR is currently stuck; I used these structures while prototyping and it's clear that they're not optimal. I already have solved most of these problems but I'm not ready yet to push everything (still need some fixes, but I am working on other Lyra stuff right now).

> First of all, calculating the term frequency should not be a concern of the tokenize: its responsibility is merely to split a string into tokens, nothing more.

100% agree.

> using a string separated by : does not sound like a good idea

knew that from the beginning, as you can see in the first comments!

> The getTokenFrequency function is also introducing an inefficiency

already rewrote it, I actually found some of the cases you described

Thanks as always for your comments, they're actually super useful and really apreciated

Reply:
Welcome :)